src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs

```diff
@@ -23,59 +23,138 @@
+                    }
+                    if (createTicketContext.Skipped) {
+                        return AuthenticateResult.Skip();
+                    }
+
+                    // Create a new authentication ticket from the introspection
+                    // response returned by the authorization server.
+                    ticket = createTicketContext.Ticket ?? await CreateTicketAsync(payload);
+                    Debug.Assert(ticket != null);
+
+                    await StoreTicketAsync(token, ticket);
+                }
+
+                // Allow for interception and handling of the token validated event.
+                var tokenValidatedContext = new TokenValidatedContext(Context, Options, ticket);
+                await Options.Events.TokenValidated(tokenValidatedContext);
+                if (tokenValidatedContext.HandledResponse)
+                {
+                    return AuthenticateResult.Success(tokenValidatedContext.Ticket);
                 }
+                if (tokenValidatedContext.Skipped)
+                {
+                    Logger.LogInformation("Authentication was skipped by event processing.");
 
-                // Ensure that the access token was issued
-                // to be used with this resource server.
-                if (!await ValidateAudienceAsync(payload)) {
-                    return AuthenticateResult.Fail("Authentication failed because the access token " +
-                                                   "was not valid for this resource server.");
+                    return AuthenticateResult.Skip();
                 }
 
-                // Create a new authentication ticket from the introspection
-                // response returned by the authorization server.
-                ticket = await CreateTicketAsync(payload);
-                Debug.Assert(ticket != null);
+                // Flow the ticket changes
+                ticket = tokenValidatedContext.Ticket;
 
-                await StoreTicketAsync(token, ticket);
-            }
+                // Ensure that the authentication ticket is still valid.
+                if (ticket.Properties.ExpiresUtc.HasValue &&
+                    ticket.Properties.ExpiresUtc.Value < Options.SystemClock.UtcNow) {
+                    return AuthenticateResult.Fail("Authentication failed because the access token was expired.");
+                }
 
-            // Ensure that the authentication ticket is still valid.
-            if (ticket.Properties.ExpiresUtc.HasValue &&
-                ticket.Properties.ExpiresUtc.Value < Options.SystemClock.UtcNow) {
-                return AuthenticateResult.Fail("Authentication failed because the access token was expired.");
+                return AuthenticateResult.Success(ticket);
             }
+            catch (Exception exception)
+            {
+                var authenticationFailedContext = new AuthenticationFailedContext(Context, Options) {
```

Comment:
> Although counting of failed authentications to limit system-abuse would be a nice-to-have thingy.

Actually, you can't use this event for that, since "invalid tokens" are rejected without throwing an exception (it just returns `AuthenticateResult.Failed`). That said, I agree it's an interesting scenario we may/should support in the future. Feel free to open a ticket to track this feature :+1: